datafusion/physical-plan/src/sorts/merge.rs

```diff
@@ -167,23 +160,28 @@ impl<C: CursorValues> SortPreservingMergeStream<C> {
         // try to initialize the loser tree
         if self.loser_tree.is_empty() {
             // Ensure all non-exhausted streams have a cursor from which rows can be pulled
-            let remaining_partitions = self.uninitiated_partitions.clone();
-            for i in remaining_partitions {
+            let mut any_pending = false;
+            for i in 0..self.streams.partitions() {
                 match self.maybe_poll_stream(cx, i) {
                     Poll::Ready(Err(e)) => {
                         self.aborted = true;
                         return Poll::Ready(Some(Err(e)));
                     }
-                    Poll::Pending => {
-                        self.uninitiated_partitions.rotate_left(1);
-                        cx.waker().wake_by_ref();
-                        return Poll::Pending;
+                    Poll::Ready(Ok(())) => {
+                        // input i is ready
                     }
-                    _ => {
-                        self.uninitiated_partitions.retain(|idx| *idx != i);
+                    Poll::Pending => {
+                        // input i is not ready
+                        any_pending = true;
                     }
                 }
             }
+            if any_pending {
+                // If any stream is not ready, return pending and tell the executor to wake us up
+                // to try again when it is ready
+                cx.waker().wake_by_ref();
```

Comment:
I have found, like @berkaysynnada, that it is required to call the waker here in order to have the sort preserving merge stream polled again. 

I am somewhat worried about the use of `wake_by_ref` as I worry it may result in a busy loop (as I think it signals to tokio to call `poll` again as soon as possible (likely after trying to poll each input stream)

BTW I also confirmed that the `ready!` macro does not invoke the wake_by_ref
https://docs.rs/futures-core/0.3.30/src/futures_core/task/poll.rs.html#5